Add checks in Shells test suite

pratikjagrut opened the pull request:
Shells.robot test suite fails because it won't find BINARY_ROOT or any other paths associated with it. So this code will add some checks before running the test and makes sure to generate related bin at the defined path.

TO test: ` ROBOT_RUN_TESTS=shells.robot make acceptance`

pratikjagrut commented:
@marckhouzam Can you review this PR?

marckhouzam commented:
The way we've been running these tests, is with an existing clone of the helm repository on our machines, which has the helm binary already built.
To specify where that binary resides, you must set ROBOT_HELM_PATH

For example I have
```
$HOME/git/acceptance-testing
$HOME/git/helm
```
So I run
`ROBOT_HELM_PATH=$HOME/git/helm/bin ROBOT_RUN_TESTS=shells.robot make acceptance` 

Can you clarify what scenario you are trying to fix with this PR?

pratikjagrut commented:
Even if we run the tests on given helm bin path, it fails because https://github.com/helm/acceptance-testing/blob/master/scripts/completion-tests/test-completion.sh#L64 because it searches helm bin in `BINARY_PATH_DOCKER=${BINARY_ROOT}/../_dist/linux-amd64` and helm bin under this path will only exist after Running 'make build-cross' in a clone of helm repo. So my code will check if helm repo exists or not if not it clones then runs `make build build-cross` to generate bin on defined locations.

marckhouzam commented:
I like the idea. Thanks.

marckhouzam commented on scripts/completion-tests/test-completion.sh:
```diff
@@ -34,10 +34,26 @@ export -f set_shell_debug_level
-BINARY_PATH_DOCKER=${BINARY_ROOT}/../_dist/linux-amd64
+HELM_DIR=${SCRIPT_DIR}/../../.acceptance
+HELM_REPO=${HELM_DIR}/helm
+BINARY_ROOT=${HELM_REPO}/bin
```
The change should continue to respect `ROBOT_HELM_PATH` if it is specified.

pratikjagrut replied:
Yes you are right it should consider `ROBOT_HELM_PATH`, but if you [see here](https://github.com/helm/acceptance-testing/blob/master/scripts/completion-tests/test-completion.sh#L38) `_dist/linux-amd64` this dir is explicitly species here so even if set helm path its going to search `_dist/linux-amd64` and if not found will fail the test.

marckhouzam commented on scripts/completion-tests/test-completion.sh:
```diff
@@ -34,10 +34,26 @@ export -f set_shell_debug_level
+        exit $exit_code;
+    fi
+else
+    (cd ${HELM_DIR} && git clone https://github.com/helm/helm.git && cd helm && make build build-cross) &> /dev/null
```
We should only clone one level of depth to clone faster

pratikjagrut replied:
yes you are right we could do that, nice suggestion, and one more issue I thought could be what if a person wants to run this test on any specified branch of the helm repo then he would have to modify this line. So instead of cloning what if we provide a direct path of bin here, anyway [this is](https://github.com/helm/acceptance-testing/blob/master/scripts/completion-tests/test-completion.sh#L38) a path of `linux-amd64/bin`. It will be a small, efficient and elegant fix and we won't need cloning repo anymore.

marckhouzam commented on scripts/completion-tests/test-completion.sh:
```diff
@@ -34,10 +34,26 @@ export -f set_shell_debug_level
 BINARY_PATH_LOCAL=${BINARY_ROOT}
 
+if [ -d ${HELM_REPO} ]; then
+    (cd ${HELM_REPO} && make build build-cross) &> /dev/null
```
It will be hard to understand what happened if the build fails. Would it make sense to let the output show on the console?

pratikjagrut replied:
I think it only sends stdout to null not stderr, so if the build fails it should output on console.

marckhouzam replied:
Normally `>` is for stdout but `&>` is for both stdout and stderr.

pratikjagrut commented:
I've updated my approach instead of cloning and generating helm bin, we can use provided bin, this way we can give the people more latitude to test whichever bin they want to test.

pratikjagrut commented:
@marckhouzam What do you think of this updated approach?

marckhouzam commented:
> I've updated my approach instead of cloning and generating helm bin, we can use provided bin, this way we can give the people more latitude to test whichever bin they want to test.

That is what $ROBOT_HELM_PATH is meant to do.
There are two helm binaries to test:
* The local binary when running on a Mac
* The Linux binary for the tests in docker

So, I believe what you are looking for is already provided with ROBOT_HELM_PATH.

pratikjagrut commented:
> > I've updated my approach instead of cloning and generating helm bin, we can use provided bin, this way we can give the people more latitude to test whichever bin they want to test.
> 
> That is what $ROBOT_HELM_PATH is meant to do.
> There are two helm binaries to test:
> 
> * The local binary when running on a Mac
> * The Linux binary for the tests in docker
> 
> So, I believe what you are looking for is already provided with ROBOT_HELM_PATH.

@marckhouzam Yes, we can provide `ROBOT_HELM_PATH` and it will be assigned to [BINARY_ROOT](https://github.com/helm/acceptance-testing/blob/master/scripts/completion-tests/test-completion.sh#L37).
Now, in [BINARY_PATH_DOCKER=${BINARY_ROOT}/../_dist/linux-amd64](https://github.com/helm/acceptance-testing/blob/master/scripts/completion-tests/test-completion.sh#L38), will try to search Helm bin in the wrong dir.

Suppose `ROBOT_HELM_PATH=/local/usr/bin` then `BINARY_ROOT` will be set to `BINARY_ROOT=/local/usr/bin` and now `BINARY_PATH_DOCKER` will be set to`BINARY_PATH_DOCKER=/local/usr/bin/../_dist/linux-amd64`. Which will be incorrect location for bin to test in docker. 

And if we don't set `ROBOT_HELM_PATH` then bin to test docker is suppose to be searched in helm repo.

So I removed this dependency of helm repo and directly running the test on the provided bin.

marckhouzam commented:
> Suppose `ROBOT_HELM_PATH=/local/usr/bin` then `BINARY_ROOT` will be set to `BINARY_ROOT=/local/usr/bin` and now `BINARY_PATH_DOCKER` will be set to`BINARY_PATH_DOCKER=/local/usr/bin/../_dist/linux-amd64`. Which will be incorrect location for bin to test in docker.

I finally get it!  That was a good example.

I assume you are running on Linux?  And things don't work unless you use a helm repo.

The problem is that BINARY_PATH_LOCAL is only used for Darwin (MacOS) so it cannot be the same as BINARY_PATH_DOCKER.  When running the tests on MacOS, two different binaries are needed.  But when running on Linux only a single binary is needed.

Maybe if we put some checks that if we are not on Darwin, we do your new logic?

pratikjagrut commented:
> I finally get it! That was a good example.

Thank you @marckhouzam.

> I assume you are running on Linux? And things don't work unless you use a helm repo.

Yes, I'm using Linux.

> The problem is that BINARY_PATH_LOCAL is only used for Darwin (MacOS) so it cannot be the same as BINARY_PATH_DOCKER. When running the tests on MacOS, two different binaries are needed. But when running on Linux only a single binary is needed.

I didn't consider the MacOs case, my bad.

> Maybe if we put some checks that if we are not on Darwin, we do your new logic?

Yes, this is definitely the way to do, we can add `if-else` or `case` construct here. Also, we should introduce a new flag, like `ROBOT_HELM_PATH_LINUX` making it compulsory to set if we are using darwin system. Now `ROBOT_HELM_PATH` may be set to darwin based bin or maybe unset and `ROBOT_HELM_PATH_LINUX` must be set to Linux bin. This way we can make sure we are using two different bin while testing on darwin and also we wouldn't need to clone helm repo and build it. WDYT?

marckhouzam commented:
> Also, we should introduce a new flag, like ROBOT_HELM_PATH_LINUX making it compulsory to set if we are using darwin system.

Actually, I think we can go simpler than that.  On Mac, there won't be a Linux binary available normally.  To test one, the user has to build one using `build-cross`. So we can assume it will be in `${BINARY_ROOT}/../_dist/linux-amd64 ` like we do now.

So I suggest that when on Mac, we check if the Linux binary is present like we do now, but if it's not, we skip the docker tests with a message to the user. That's a nicer improvement for me as it will be simple to just run the Mac tests during development which is much faster.

pratikjagrut commented:
> Actually, I think we can go simpler than that. On Mac, there won't be a Linux binary available normally. To test one, the user has to build one using `build-cross`. So we can assume it will be in `${BINARY_ROOT}/../_dist/linux-amd64 ` like we do now.

What if a developer or a tester using macOS and wants to test both(Linux based and darwin based) `external binaries`? He can set only one binary path using `ROBOT_HELM_PATH`. In this, we can not assume that Linux binary will be under `${BINARY_ROOT}/../_dist/linux-amd64`.

marckhouzam commented:
> What if a developer or a tester using macOS and wants to test both(Linux based and darwin based) `external binaries`? 

How would that person get both binaries?  I was assuming they would have to build it.  Would they download both binaries?  It is not clear to me what use case it is where someone would want to run the acceptance-tests on a binary that they didn't modify themselves...

I don't want to overly complicate things with extra variables unless there is a specific need.

pratikjagrut commented:
I think I'm puzzled by my own scenario. I'll first implement as we discussed here  
> Actually, I think we can go simpler than that. On Mac, there won't be a Linux binary available normally. To test one, the user has to build one using `build-cross`. So we can assume it will be in `${BINARY_ROOT}/../_dist/linux-amd64 ` like we do now.

and I'll see if this satiate my use case and if not then probably I'll be in better condition to explain my use case.

pratikjagrut commented:
Hey @marckhouzam, I updated code as you suggested please review it.

> How would that person get both binaries? I was assuming they would have to build it. Would they download both binaries? It is not clear to me what use case it is where someone would want to run the acceptance-tests on a binary that they didn't modify themselves...

What I was trying to suggest before was, suppose if I have a helm downstream repo and I generated a helm Linux based bin and sent it to my QA for testing who uses macOS. So now how he will test?
 
He has a choice of explicitly copying that bin under `helm/_dist/Linux-amd64` which will be consumed by tests running in docker. Or we can introduce a flag which will be set to linux based bin while running tests on macOS or any other OS and this copying can be avoided.

WDYT?

pratikjagrut commented:
Hey, @marckhouzam Can you review updated approach?

marckhouzam commented:
Sorry for the delay @pratikjagrut, I keep having trouble with this PR because I keep trying to think of the valid use cases we need to cover.  How about we list the use cases, then we can make sure they are covered.  Here are the ones I can think of:

1. Linux platform
    - only docker tests should be run, no local tests
    - the user may want to test helm that is on `$PATH`,
 OR
    - the user may specify a helm to test using `$ROBOT_HELM_PATH`
1. MacOS platform
    - tests in docker (linux build) and local tests (darwin build)
    - which linux build to use? now we hard-code to `linux-amd64` but maybe we should check the machines architecture
    - if the user wants to test helm that is on `$PATH`, I think we should only run local tests as the helm on `$PATH` would logically be build for the current MacOS plateform
    - if the user specifies `$ROBOT_HELM_PATH` then we run local tests with `$ROBOT_HELM_PATH/helm`, but how do we find the linux build for the docker tests?

Does this cover the cases you were thinking of?

pratikjagrut commented:
> 1. Linux platform
>    
>    * only docker tests should be run, no local tests
>    * the user may want to test helm that is on `$PATH`,
>      OR
>    * the user may specify a helm to test using `$ROBOT_HELM_PATH`

I comply with Linux use cases.

> 2. MacOS platform
>    
>    * tests in docker (linux build) and local tests (darwin build)
>    * which linux build to use? now we hard-code to `linux-amd64` but maybe we should check the machines architecture

I think we won't need to check machines architecture, as the docker images are hardcoded and probably they are for amd64.

>    * if the user wants to test helm that is on `$PATH`, I think we should only run local tests as the helm on `$PATH` would logically be build for the current MacOS plateform

Agree with this usecase.

>    * if the user specifies `$ROBOT_HELM_PATH` then we run local tests with `$ROBOT_HELM_PATH/helm`, but how do we find the linux build for the docker tests?

To find Linux build we can provide new flag `$ROBOT_HELM_LINUX_PATH` for macOS. 
OR
We've can skip the docker tests for the specified binaries same as we do for `$PATH`.
I would like to go with the first solution unless you feel it's gratuitous then we've to go with the second one or if you got any better solution.

> 
> Does this cover the cases you were thinking of?

Yes, these are my current use cases.

pratikjagrut commented:
@marckhouzam WDYT of above comment?

marckhouzam commented:
> > 1. MacOS platform
> >    
> >    * tests in docker (linux build) and local tests (darwin build)
> >    * which linux build to use? now we hard-code to `linux-amd64` but maybe we should check the machines architecture
> 
> I think we won't need to check machines architecture, as the docker images are hardcoded and probably they are for amd64.

ok for now.

> > * if the user specifies `$ROBOT_HELM_PATH` then we run local tests with `$ROBOT_HELM_PATH/helm`, but how do we find the linux build for the docker tests?
> 
> To find Linux build we can provide new flag `$ROBOT_HELM_LINUX_PATH` for macOS.
> OR
> We've can skip the docker tests for the specified binaries same as we do for `$PATH`.
> I would like to go with the first solution unless you feel it's gratuitous then we've to go with the second one or if you got any better solution.

Well, looking at your code changes, for docker tests on MacOS, the user would need to put the binary under `${SCRIPT_DIR}/../../../helm/_dist/linux-amd64`.
That is good enough for me at this time.

So I'm ok with the PR as it is right now.
Could you do a little cleanup please, and then I will merge:
1- squash your commits into one with a single commit message
2- do a `git diff` or `git show` and remove the trailing whitespaces it will show you.

Thank you for your patience.

pratikjagrut commented:
Cool @marckhouzam I'll do the cleanup.

pratikjagrut commented:
@marckhouzam I pushed squashed commit, have a look.

marckhouzam commented:
> @marckhouzam I pushed squashed commit, have a look.

Thank you @pratikjagrut!
